### DISCUSSION TITLE
Why is voltage limited to 4200?

### DISCUSSION THREAD

**Discussion by llucax** [VR-25/acc#88] [Q&A] [answered]
Hi, I have a LG G6 and `cat /sys/class/power_supply/battery/voltage_max_design` returns `4350000` (it seems to have some wrong extra `000`), but the point is 4350 it's bigger than the maximum limit imposed by acc (of 4200). Is there any reason why not using `voltage_max_design` instead of hardcoding it for all devices to 4200?

As a side question, this device is supposed to support voltage up to 9V (and 1.8amp) in fast charge mode. Also my understanding is that 5V is the maximum voltage that an USB port normally gives, so I wonder why the limits are so small when charging.

PS: I know using the maximum voltage is bad for the battery.

**Comment by VR-25** [comment, marked as answer]
`voltage_max_design`, as the name implies, refers to the device's maximum _charging_ voltage (not to be confused with input voltage - more on that later).
The _extra zeroes_ mean the unit is microvolt.
I have never seen volt being used.
It's usually either millivolt (e.g., 4350) or microvolt (4350000), for greater control/accuracy.

Lithium-based batteries are very sensitive to voltage and cannot tolerate overcharging at all - unlike, for instance, lead-acid batteries.
The former carry much higher risk of fire/explosion or rapid degradation when exposed to excessive voltage.

The 4200 mV limit imposed by acc is a safety measure.
It takes legacy batteries into account.
Originally, lithium-based batteries were charged to 4200 mV max.

Now let's talk about input vs charging powers.
Modern batteries can charge up to 4.4 volts.
Input power is converted to accommodate the battery's max voltage and current limits.

ACC does not limit what the power brick delivers.
Rather, it limits what actually goes to the battery - the already converted power.
You can override the 4200 mV max limit by manually editing the config file.

**Reply by llucax** [reply]
Thanks for the detailed explanation!

And any ideas about the second part? The phone supports fast charge, and after using ACC I have the feeling that fast charging is not working as fast as it did. But then I see that the `voltage_max_design` is not even close to 9V and I guess ACC could have never changed that.

**Reply by VR-25** [reply]
Rephrasing, `voltage_max_design` is the maximum _charging_ voltage limit set by the device's manufacturer.
Keep in mind that it differs from _input_ voltage.
The former won't/cannot be higher than 4.4 Volts, while the latter can be 5, 9 or more Volts.

voltage_max_design is strictly read-only - meaning, acc cannot modify it at all.
The target is a different file - and it comes in different names (e.g., voltage_max), depending on the device.

Limiting the charging voltage to a value below voltage_max_design also limits fast charging to some degree.
Besides, it always limits the available capacity - i.e., the battery won't charge fully.

In sum, voltage_max_design is 4.4V max, never 5V, let alone 9V.
Those higher numbers refer to input, as opposed to charging voltage.
Input voltage and current are converted to values the device and battery can use safely - e.g., protection circuits aside, 9V would fry the phone and/or explode the battery.

P.S., limiting the charging current (not voltage) to a modest value (e.g., 1A) prevents fast charging).
You don't want this if your goal is faster charging.

Side Note
The input voltage can generally be read from `/sys/class/power_supply/usb/voltage_max`.
That's where you might see 9V, althogh with _extra zeroes_ (finer unit).

**Reply by llucax** [reply]
Thanks again for the detailed info! Also, please just say so if I'm abusing and taking this too far, I totally get it if that's the case :)

But if not, I'm curious...

I believe some devices implement fast charging by providing high current with the standard 5V and some (like the LG G6) high voltage too (9V in this case). Why is high voltage used if the battery can only get 4.2V for charging? Is it to be able to support the device normal functioning via the power supply without impacting the charging speed even if the device is in need of a lot of power while charging?

Looking at my device's `/sys/class/power_supply/battery` files:

```console
lucye:/sys/class/power_supply/battery # grep -n . *current* *flash* | cut -d: -f1,3                                                                                
constant_charge_current:-22000
constant_charge_current_max:800000
current_max:500000
current_now:-240629
flash_current_max:1310446
input_current_limit:4096
input_current_limited:1
input_current_max:1700000
input_current_settled:1
flash_active:0
flash_current_max:1448272
flash_trigger:0
```

This is connected to an USB hub that delivers very little power. I guess all values are in µA and negative values means the current is being delivered to the battery instead of consumed from the battery, correct?

Do you know what all these values are about? Is `input_current*` what the device can get from the power supply and `current_*` is what the battery takes? Is `flash_*` about fast charging? And `constant_charge_current*`?

Also, is it normal not to have any `max_design` with current as there is one for `voltage`, telling the device's design maximum?

Thanks!

**Reply by VR-25** [reply]
> I believe some devices implement fast charging by providing high current with the standard 5V and some (like the LG G6) high voltage too (9V in this case). Why is high voltage used if the battery can only get 4.2V for charging? Is it to be able to support the device normal functioning via the power supply without impacting the charging speed even if the device is in need of a lot of power while charging?

First, recall that input power conversions take place, as we discussed previously.
Now, simply put, high voltage is used for efficiency reasons.
Achieving fast charging with lower voltage means there must be higher current passing through the wires.
This would require thicker wires (lower [AWG](https://en.wikipedia.org/wiki/American_Wire_Gauge)).
Those take up more space, are heavier and more expensive.
For additional data, you can refer to the history of electricity - Nikola Tesla's "Alternating Current" vs Thomas Edison's "Direct Current").
Related: [Why is electricity transmitted at high voltages?](https://www.electricalclassroom.com/why-is-electricity-transmitted-at-high-voltages/)

---
> /sys/class/power_supply/battery/

Unfortunately, this interface is not standardized.
Thus, I cannot answer your questions with great certainty without studying that particular device a bit deeper.
Manufacturers like to do their own mess here.

---
Those pairs mean the same things and  are supposed to have the same values.
The former pair refers to real time charging current readings.
A negative sign may either mean _discharge_ or nothing at all (depends on device).
The latter pair indicates the charging current limit.
> constant_charge_current:-22000
current_now:-240629

> constant_charge_current_max:800000
current_max:500000

---
`input_current_limit:4096` (value/unit is weird, compared to `input_current_max` below)
`input_current_limited:1` (whether `input_current_limit` should be honored)
`input_current_max:1700000` (uA, 1.7A, seems to be read-only)
`input_current_settled:1` (?)

---
These may be related to laser beams.
Your phone may be part of some alien ship weaponry.
> flash_current_max:1310446
flash_active:0
flash_current_max:1448272
flash_trigger:0

---
Unlike voltage, we don't have to worry about current off-limits.
Higher than supported current values are simply ignored.
`input_current_max` seems to be the equivalent of `voltage_max_design`.

---
Again, to understand all the weird conventions each manufacturer uses, one has to read the drivers'/kernel documentations.
Oftentimes, that's not enough.
It takes some deeper digging and testing to understand the non-intuitive stuff.
Just for fun, try to make sense of what's in a MediaTek device's /sys/class/power_supply/.

P.S., great questions, feel free to keep them coming!
Although I don't have much free time, I try to answer them all to the best of my knowledge, at my earliest convenience.

**Reply by llucax** [reply]
Thanks once more for taking the time to have a look at this and answer my questions!

I think I'm good for now, but is good to know you enjoy answering in case I have some more :)

This was very informative!
